src/database.rs

```diff
@@ -3,761 +3,788 @@ use argon2::{
-        .execute(conn)?;
-    Ok(res != 0)
-}
+    fn are_friends(&mut self, user: i32, friend_id: i32) -> Result<bool, TimeError> {
+        use crate::schema::friend_relations::dsl::*;
+        let (lesser, greater) = if user < friend_id {
+            (user, friend_id)
+        } else {
+            (friend_id, user)
+        };
+        Ok(friend_relations
+            .filter(lesser_id.eq(lesser).and(greater_id.eq(greater)))
+            .first::<FriendRelation>(self)
+            .optional()?
+            .is_some())
+    }
 
-pub fn promote_user_to_leaderboard_admin(
-    conn: &mut PooledConnection<ConnectionManager<PgConnection>>,
-    lid: i32,
-    uid: i32,
-) -> Result<bool, TimeError> {
-    use crate::schema::leaderboard_members::dsl::*;
-    let res = diesel::update(crate::schema::leaderboard_members::table)
-        .filter(user_id.eq(uid).and(leaderboard_id.eq(lid)))
-        .set(admin.eq(true))
-        .execute(conn)?;
-    Ok(res != 0)
-}
+    fn remove_friend(&mut self, user: i32, friend_id: i32) -> Result<bool, TimeError> {
+        use crate::schema::friend_relations::dsl::*;
+        let (lesser, greater) = if user < friend_id {
+            (user, friend_id)
+        } else {
+            (friend_id, user)
+        };
+        Ok(diesel::delete(friend_relations)
+            .filter(lesser_id.eq(lesser).and(greater_id.eq(greater)))
+            .execute(self)?
+            != 0)
+    }
 
-pub fn demote_user_to_leaderboard_member(
-    conn: &mut PooledConnection<ConnectionManager<PgConnection>>,
-    lid: i32,
-    uid: i32,
-) -> Result<bool, TimeError> {
-    use crate::schema::leaderboard_members::dsl::*;
-    let res = diesel::update(crate::schema::leaderboard_members::table)
-        .filter(user_id.eq(uid).and(leaderboard_id.eq(lid)))
-        .set(admin.eq(false))
-        .execute(conn)?;
-    Ok(res != 0)
-}
+    fn regenerate_friend_code(&mut self, userid: i32) -> Result<String, TimeError> {
+        use crate::schema::user_identities::dsl::*;
+        let code = crate::utils::generate_friend_code();
+        diesel::update(crate::schema::user_identities::table)
+            .filter(id.eq(userid))
+            .set(friend_code.eq(&code))
+            .execute(self)?;
+        Ok(code)
+    }
 
-pub fn is_leaderboard_member(
-    conn: &mut PooledConnection<ConnectionManager<PgConnection>>,
-    uid: i32,
-    lid: i32,
-) -> Result<bool, TimeError> {
-    use crate::schema::leaderboard_members::dsl::*;
-    Ok(leaderboard_members
-        .filter(user_id.eq(uid).and(leaderboard_id.eq(lid)))
-        .select(id)
-        .first::<i32>(conn)
-        .optional()?
-        .is_some())
-}
+    fn delete_activity(&mut self, userid: i32, activity: i32) -> Result<bool, TimeError> {
+        use crate::schema::coding_activities::dsl::*;
+        let res = diesel::delete(crate::schema::coding_activities::table)
+            .filter(id.eq(activity))
+            .filter(user_id.eq(userid))
```

Comment:
Aren't activity ids unique? If yes, then why are we using `user_id` here?

Reply:
They are unique, but this ensure a user cant delete other user's activities